### ISSUE TITLE
amp-analytics option select not triggering request

### ISSUE THREAD

**Issue by RonnieWinter** [ampproject/amphtml#17878]
## What's the issue?

When trying to use amp-analytics to track a selection in an option select using the data-vars-* attribute, I notice that the event does not get tracked by amp-analytics.

Please refer to this [CodePen](https://codepen.io/ro2nie/pen/YOxPyy) to see this happening.

As you can see in the CodePen example, there is an <amp-analytics> element to track the click on the <select> which successfully triggers. But then the second <amp-analytics> element, which is concerned with tracking the actual <option> value, never fires.
You can see this on the network tab in the dev tools.

Is there any better way of tracking an option select with amp-analytics that I could use instead of this method?

**Comment by zhouyx** [comment]
@ro2ni3 I am able to reproduce this. The reason behind is that the `<option>` doesn't fire click event when been clicked, but only `onchange` event. 

Few ideas I had. @aghassemi @cvializ
1. Have the form listen to the onchange event and fire a customized analytics event. Would that be possible?
2. Introduce a new `onchange` trigger.
3. Is it possible that we can implement the `<select><option></option></select>` with some AMP components?

**Comment by aghassemi** [comment]
yeah `select` does not fire `onclick` (rightly so since there are different UIs to chose an option. e.g. the wheel on iOS).

`amp-selector` can be used to mimic `select` to some extend but native is always preferred when possible.

`onchange` trigger seems reasonable.

**Comment by zhouyx** [comment]
After talking to the team, we decide it is reasonable to introduce a customizable trigger that allows tracking of events like `onchange` or `onblur`..
cc @lannka @jeffjose

**Comment by stephengardner** [comment]
Hey gang, been a while.  Has there been any progress on this front?

**Comment by zhouyx** [comment]
Hello @stephengardner. Thank you for bumping this up. Unfortunately the analytics working group hasn't got a chance to work on this feature request. 
We'll update here if there's any progress. In the meanwhile, any input or contribution is welcome.

**Comment by kalemuw** [comment]
Hello @choumx @stephengardner @RonnieWinter. I am currently working on a customizable trigger list that will serve as a fix for the issue discussed in this thread. 
Below are some of the implementation details. I would appreciate any feedback. 


<b style="font-weight:normal;" id="docs-internal-guid-390728e8-7fff-da5a-4745-49919243b1a1">
<h1 dir="ltr" style="line-height:1.38;margin-top:24pt;margin-bottom:4pt;">
<span style="font-size:18pt;font-family:'Open Sans',sans-serif;color:#1976d2;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">
Overview
</span>
</h1>
<p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;">
<span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">
The aim of this feature is to develop an API to track a customizable list of events that are not yet supported by amp-analytics. 
&nbsp;
</span>
</p>
<p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;">
<span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">
The API will have the following high level structure&nbsp;
</span>
</p>
<ol style="margin-top:0;margin-bottom:0;padding-inline-start:48px;">
<li dir="ltr" style="list-style-type:upper-roman;font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">AllowList of browser events to be tracked</span></p></li><li dir="ltr" style="list-style-type:upper-roman;font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Configuration : A config that supports tracking the defined set of events on a single or a set of elements.&nbsp;</span></p></li></ol><div dir="ltr" style="margin-left:0pt;" align="left">

```
genericTrigger:{
   "on": "browser-event",
    "request": "event",
     "browserEvent":"blur",
     "selector": ".input-container",
     "vars": {
            "eventType": "blur"
      }
}
```

</div><ol style="margin-top:0;margin-bottom:0;padding-inline-start:48px;" start="3"><li dir="ltr" style="list-style-type:upper-roman;font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Custom Browser Events Tracker</span></p></li></ol><p dir="ltr" style="line-height:1.38;margin-left: 36pt;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The Custom Browser Event Tracker will implement a tracker on events defined in the AllowList using the publisher defined selectors. This trigger will then fire a request.&nbsp;</span></p></b><br class="Apple-interchange-newline">

<b style="font-weight:normal;" id="docs-internal-guid-3dbc9c61-7fff-63e1-a032-fe27695b083c"><h1 dir="ltr" style="line-height:1.38;margin-top:24pt;margin-bottom:4pt;"><span style="font-size:18pt;font-family:'Open Sans',sans-serif;color:#1976d2;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Detailed design</span></h1><ol style="margin-top:0;margin-bottom:0;padding-inline-start:48px;"><li dir="ltr" style="list-style-type:upper-roman;font-size:11pt;font-family:Roboto,sans-serif;color:#1c4587;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:10pt;" role="presentation"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#1c4587;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Configuration</span></p></li></ol><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:10pt;"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">We will use the configuration to extract the trigger type, selectors for the elements to be tracked and specs that may be used in future implementations. Below is a sample configuration:</span></p><div dir="ltr" style="margin-left:0pt;" align="left">
```
"genericTrigger-1":{
            "on": "browser-event",
            "request": "event",
            "browserEvent":"blur",
            "selector": ".input-container",
            "vars": {
              "eventType": "blur"
            }
            "spec": {}
          },
   "genericTrigger-2":{
            "on": "browser-event",
            "request": "event",
            "browserEvent":"change",
            "selector": "option[data-vars-num-child]",
            "vars": {
                        "eventType": "change"
                  }
            "spec": {}
          }

```
</div><br /><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:10pt;"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The config definition for custom browser events are similar to the config of other trigger configurations</span><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">. The main features of the config are the event type, selector, and spec.&nbsp;</span></p><ul style="margin-top:0;margin-bottom:0;padding-inline-start:48px;"><li dir="ltr" style="list-style-type:disc;font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The event type is defined by the </span><span style="font-size:11pt;font-family:Consolas,sans-serif;color:#0d904f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">&ldquo;on&rdquo;</span><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;"> value and specifies the general browser event the configuration tracks.&nbsp;</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The specific browser event is indicated by the </span><span style="font-size:10pt;font-family:Consolas,sans-serif;color:#0f9d58;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">&quot;browserEvent&quot;</span><span style="font-size:10pt;font-family:Consolas,sans-serif;color:#616161;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;"> </span><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">value. For instance, events blur and change are defined in the config above.&nbsp;</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" role="presentation"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The selector value lists the elements on the page on which the specified custom browser event will be tracked on. The value can be a single string or an array of strings of valid CSS query selectors.&nbsp;&nbsp;</span></p></li><li dir="ltr" style="list-style-type:disc;font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;" aria-level="1"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:10pt;" role="presentation"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">The spec value contains properties of the events.&nbsp;</span></p></li></ul><p dir="ltr" style="line-height:1.38;margin-left: 36pt;margin-top:0pt;margin-bottom:10pt;"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Currently the spec value is empty due to the fact that there aren&rsquo;t any implementation needs. However, as other event trackers have shown, it may become useful in future developments.&nbsp;</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:10pt;"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#1976d2;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Special Selector Values&nbsp;</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:10pt;"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">As stated above the selector value can be a single string of CSS selector or an array of string selectors. In other triggers special selectors suchs as :host, :ini-load, and :root are supported. However, for the custom browser events that we are supporting on initial development (on-blur and on-change), such selectors have little developmental use and thus, we see no need to support them at current time.&nbsp;</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:10pt;"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#1976d2;background-color:transparent;font-weight:400;font-style:italic;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Selection Method</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:10pt;"><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Selection Method is a special key that can be defined in the configuration to change the semantics of how elements are matched. It can have one of two values, </span><span style="font-size:11pt;font-family:Consolas,sans-serif;color:#0d904f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">scope</span><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;"> or </span><span style="font-size:11pt;font-family:Consolas,sans-serif;color:#0d904f;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">closest</span><span style="font-size:11pt;font-family:Roboto,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">. This will be supported in our implementation.</span></p></b><br class="Apple-interchange-newline">

**Comment by kalemuw** [comment]
Closed in #35193
